File: python/toolbox/DecayFitNetToolbox.py
import torch
import numpy as np
import pickle
import onnx
import onnxruntime
import os
from pathlib import Path
from typing import List, Tuple

from .core import PreprocessRIR, _postprocess_parameters


class DecayFitNetToolbox:
    def __init__(self, n_slopes: int = 0, sample_rate: int = 48000, filter_frequencies: List = None,
                 device: torch.device = torch.device('cpu')):
        self._version = '0.1.2'
        self.device = device

        self._n_slopes = n_slopes
        self._sample_rate = sample_rate

        # setup path
        path_onnx = Path.joinpath(Path(__file__).parent.parent.parent, 'model')

        # load net and start onnx session
        if n_slopes == 0:
            n_slope_str = ''
        else:
            n_slope_str = f'_{n_slopes}slopes'
        model_name = f'DecayFitNet{n_slope_str}_v10.onnx'
        self._onnx_model = onnx.load(os.path.join(path_onnx, model_name))
        onnx.checker.check_model(self._onnx_model)
        self._session = onnxruntime.InferenceSession(os.path.join(path_onnx, model_name))

        # Load input transform
        self._input_transform = pickle.load(open(os.path.join(path_onnx, f'input_transform{n_slope_str}.pkl'), 'rb'))

        self._preprocess = PreprocessRIR(input_transform=self._input_transform,
                                         sample_rate=self._sample_rate,
                                         output_size=100,
                                         filter_frequencies=filter_frequencies)

    def __repr__(self):
        frmt = f'DecayFitNetToolbox {self._version}  \n'
        frmt += f'Input fs = {self._sample_rate} \n'
        frmt += f'Output_size = {self.get_output_size()} \n'

        return frmt

    @staticmethod
    def _to_numpy(tensor):
        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

    def set_n_slopes(self, n_slopes):
        assert n_slopes <= 3, 'Maximum number of supported slopes is 3.'
        self._n_slopes = n_slopes

    def get_max_n_slopes(self):
        if self._n_slopes == 0:
            max_n_slopes = 3
        else:
            max_n_slopes = self._n_slopes
        return max_n_slopes

    def set_filter_frequencies(self, filter_frequencies):
        self._preprocess.set_filter_frequencies(filter_frequencies)

    def get_filter_frequencies(self):
        return self._preprocess.get_filter_frequencies()

    def set_output_size(self, output_size):
        self._preprocess.output_size = output_size

    def get_output_size(self):
        return self._preprocess.output_size

    def preprocess(self, signal: torch.Tensor, analyse_full_rir=True) -> Tuple[torch.Tensor, torch.Tensor]:
        """Preprocess an input signal to extract EDCs"""
        edcs, norm_vals = self._preprocess(signal, analyse_full_rir)
        return edcs, norm_vals

    def estimate_parameters(self, input: torch.Tensor, input_is_edc: bool = False, analyse_full_rir=True) -> Tuple[List[torch.Tensor], torch.Tensor]:
        """ Estimates the parameters for this impulse response. The resulting fitted EDC is normalized to 0dB and can be
        re-normalized to the original level with norm_vals

        Args:
            input: [rir_length, 1], rir to be analyzed
            input_is_edc: bool that indicates if input is edc or rir

        The estimation returns:
            Tuple of:
            1) List with:
            -- t_prediction : [batch, 3] : time_values for the 3 slopes
            -- a_prediction : [batch, 3] : amplitude_values for the 3 slopes
            -- n_prediction :  [batch, 1] : noise floor
            2) norm_vals : [batch, n_bands] : EDCs are normalized to 0dB, as customary for most decay analysis problems,
                                              but if the initial level is required, norm_vals will return it

        """
        edcs, __, norm_vals, scale_adjust_factors = self._preprocess(input, input_is_edc, analyse_full_rir)

        ort_inputs = {self._session.get_inputs()[0].name: DecayFitNetToolbox._to_numpy(edcs)}
        ort_outs = self._session.run(None, ort_inputs)

        t_prediction, a_prediction, n_exp_prediction, n_slopes_probabilities = ort_outs

        # If number of active slopes is estimated by network: Get a binary mask to only use the number of slopes
        # that were predicted, zero others
        exactly_n_slopes_mode = (self._n_slopes != 0)
        if not exactly_n_slopes_mode:
            n_slopes_prediction = np.argmax(n_slopes_probabilities, 1)
            n_slopes_prediction += 1  # because python starts at 0
            temp = np.tile(np.linspace(1, 3, 3, dtype=np.uint8), (n_slopes_prediction.shape[0], 1))
            mask = (np.tile(np.expand_dims(n_slopes_prediction, 1), (1, 3)) < temp)
            a_prediction[mask] = 0

        # Clamp noise to reasonable values to avoid numerical problems
        n_exp_prediction = np.clip(n_exp_prediction, -32, 32)
        # Go from noise exponent to noise value
        n_prediction = np.power(10, n_exp_prediction)

        t_prediction, a_prediction, n_prediction = _postprocess_parameters(t_prediction, a_prediction, n_prediction,
                                                                           scale_adjust_factors, exactly_n_slopes_mode)

        return [t_prediction, a_prediction, n_prediction], norm_vals

File: matlab/DecayFitNetToolbox.m
classdef DecayFitNetToolbox < handle
    
    properties        
    end
    properties (SetAccess = private)
        version = '0.1.2'
        outputSize = 100  % Timesteps of resampled RIRs
        nSlopes
        onnxPath
        onnxModel
        networkName
        inputTransform
        preprocessing
        sampleRate
    end
    
    methods
        function obj = DecayFitNetToolbox(nSlopes, sampleRate, filterFrequencies)
            if nargin < 1
                nSlopes = 0; % estimate number of slopes from data
            end
            if nargin < 2
                sampleRate = 48000;
            end
            if nargin < 3
                filterFrequencies = [];
            end
                
            obj.sampleRate = sampleRate;
            obj.nSlopes = nSlopes;
            
            % Load onnx model
            [thisDir, ~, ~] = fileparts(mfilename('fullpath'));
            obj.onnxPath = fullfile(thisDir, '..', 'model');
            addpath(obj.onnxPath);
            
            if obj.nSlopes == 0
                % infer number of slopes with network
                slopeMode = '';
            elseif obj.nSlopes == 1 || obj.nSlopes == 2 || obj.nSlopes == 3
                % fit exactly n slopes plus noise
                slopeMode = sprintf('%dslopes_', obj.nSlopes);
            else
                error('Please specify a valid number of slopes to be predicted by the network (nSlopes=1,2,3 for 1,2,3 slopes plus noise, respectively, or nSlopes=0 to let the network infer the number of slopes [max 3 slopes]).');
            end
            
            obj.networkName = sprintf('DecayFitNet_%s', slopeMode);
            
            pe = pyenv;
            if strcmp(pe.Version, "") || isempty(pe.Version)
                error(sprintf("MATLAB does not know where your Python executable is located. \n Please run pyenv('Version', <path_to_your_python_executable>). \n\n More info: https://www.mathworks.com/help/releases/R2022a/matlab/matlab_external/install-supported-python-implementation.html"));
            end

            % Load ONNX model:
            if exist(fullfile(obj.onnxPath, [obj.networkName, 'model.mat']), 'file')
                fprintf('Loading precompiled model %smodel.mat\n', obj.networkName)
                obj.onnxModel = load(fullfile(obj.onnxPath, [obj.networkName, 'model.mat'])).tmp;
            else
                obj.onnxModel = importONNXFunction(fullfile(obj.onnxPath, [obj.networkName, 'v9.onnx']), fullfile(obj.onnxPath, [obj.networkName, 'model']));
                tmp = obj.onnxModel;
                save(fullfile(obj.onnxPath, [obj.networkName, 'model.mat']), 'tmp');
            end
            [~, msgid] = lastwarn;
            if strcmp(msgid, 'MATLAB:load:cannotInstantiateLoadedVariable')
                error('Could not load the ONNX model. Please make sure you are running MATLAB 2020b or later and installed the following toolbox to load ONNX models in MATLAB: https://se.mathworks.com/matlabcentral/fileexchange/67296-deep-learning-toolbox-converter-for-onnx-model-format');
            end
            
            disp(obj.onnxModel)
            
            % Load input transform for preprocessing the network inputs
            fid = py.open(fullfile(obj.onnxPath, sprintf('input_transform_%sp2.pkl', slopeMode)),'rb');
            obj.inputTransform = py.pickle.load(fid);
            
            % Init preprocessing
            obj.preprocessing = PreprocessRIR(obj.inputTransform, sampleRate, filterFrequencies, obj.outputSize);
        end
                        
        function setFilterFrequencies(obj, filterFrequencies)
            obj.preprocessing.filterFrequencies = filterFrequencies;
        end
        
        function filterFrequencies = getFilterFrequencies(obj)
            filterFrequencies = obj.preprocessing.filterFrequencies;
        end
        
        function [tPrediction, aPrediction, nPrediction, normVals] = estimateParameters(obj, input, inputIsEDC)
            % Output is [nBands, nSlopes]
            
            if ~exist('inputIsEDC', 'var')
                inputIsEDC = false;
            end
            
            [edcs, ~, normVals, scaleAdjustFactors] = obj.preprocessing.preprocess(input, inputIsEDC);
            
            % Forward pass of the DecayFitNet
            net = str2func([obj.networkName, 'model']);
            [tPrediction, aPrediction, nExpPrediction, nSlopesProbabilities] = net(edcs, obj.onnxModel);
            
            % Transpose results to make them [batchSize, nSlopes]
            tPrediction = tPrediction.';
            aPrediction = aPrediction.';
            nExpPrediction = nExpPrediction.';
            nSlopesProbabilities = nSlopesProbabilities.';
            
            % Clamp noise to reasonable values to avoid numerical problems
            nExpPrediction = min(max(nExpPrediction, -32), 32);

            % Go from noise exponent to noise value
            nPrediction = 10 .^ nExpPrediction;

            % In nSlope inference mode: Get a binary mask to only use the number of slopes that were predicted, zero others
            nSlopeEstimationMode = (obj.nSlopes == 0);
            if nSlopeEstimationMode
                [~, nSlopesPrediction] = max(nSlopesProbabilities, [], 2);
                tmp = repmat(linspace(1,3,3), [length(nSlopesPrediction), 1]);
                mask = tmp <= repmat(nSlopesPrediction, [1, 3]);
                aPrediction(~mask) = 0;
            end
            
            % Postprocess parameters and also convert to original
            % time-scale
            [tPrediction, aPrediction, nPrediction] = postprocessDecayParameters(tPrediction, aPrediction, nPrediction, scaleAdjustFactors, nSlopeEstimationMode);
        end
    end
    
    methods(Static)            
        function output = discardLast5(signal)
            % Discard last 5% samples of a signal
            assert(size(signal,1) > size(signal,2), 'The signal should be in [timesteps, channels]')
            last5 = round(0.95 * size(signal,1));
            
            output = zeros(last5, size(signal,2));
            for channel = 1:size(signal,2)
                tmp = signal(:,channel);
                tmp(last5+1:end) = [];
                output(:,channel) = tmp;
            end
        end
    end
    
end
